pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,244 @@
+}
+
+// Store is responsible for getting and updating the current certificate.
+// Depending on the concrete implementation, the backing store for this
+// behavior may vary.
+type Store interface {
+	Current() (*tls.Certificate, error)
+	// Accepts the PEM data for the cert/key pair and makes the new cert/key
+	// pair the 'current' pair, that will be returned by future calls to
+	// Current().
+	Update(cert, key []byte) (*tls.Certificate, error)
+}
+
+type manager struct {
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface
+	template                 *x509.CertificateRequest
+	usages                   []certificates.KeyUsage
+	certStore                Store
+	certAccessLock           sync.RWMutex
+	cert                     *tls.Certificate
+	shouldRotatePercent      int32
+}
+
+// NewManager returns a new certificate manager. A certificate manager is
+// responsible for being the authoritative source of certificates in the
+// Kubelet and handling updates due to rotation.
+func NewManager(
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface,
+	template *x509.CertificateRequest,
+	usages []certificates.KeyUsage,
+	certificateStore Store,
+	certRotationPercent int32) (Manager, error) {
+
+	cert, err := certificateStore.Current()
+	if err != nil {
+		return nil, err
+	}
+
+	if certRotationPercent > 100 {
+		certRotationPercent = 100
+	}
+
+	m := manager{
+		certSigningRequestClient: certSigningRequestClient,
+		template:                 template,
+		usages:                   usages,
+		certStore:                certificateStore,
+		cert:                     cert,
+		shouldRotatePercent:      certRotationPercent,
+	}
+
+	return &m, nil
+}
+
+// GetCertificate returns the certificate that should be used TLS connections.
+// The value returned by this function will change over time as the certificate
+// is rotated. If a reference to this method is passed directly into the TLS
+// options for a connection, certificate rotation will be handled correctly by
+// the underlying go libraries.
+//
+//    tlsOptions := &server.TLSOptions{
+//        ...
+//        GetCertificate: certificateManager.GetCertificate
+//        ...
+//    }
+//
+func (m *manager) GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	m.certAccessLock.RLock()
+	defer m.certAccessLock.RUnlock()
+	return m.cert, nil
+}
+
+// Start will start the background work of rotating the certificates.
+func (m *manager) Start() {
+	if m.shouldRotatePercent < 1 {
+		glog.V(2).Infof("Certificate rotation is not enabled.")
+		return
+	}
+
+	// Certificate rotation depends on access to the API server certificate
+	// signing API, so don't start the certificate manager if we don't have a
+	// client. This will happen on the master, where the kubelet is responsible
+	// for bootstrapping the pods of the master components.
+	if m.certSigningRequestClient == nil {
```

Comment:
I don't think this is an error. If there is no client, there is no rotation, but as long as there is a certificate available on the file system, the certificate manager still functions.